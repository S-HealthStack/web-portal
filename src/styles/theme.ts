import { css, DefaultTheme } from 'styled-components';
import { px } from './utils';

const colors = {
  // gray gradations
  black: '#000000',
  black40: 'rgba(0, 0, 0, 0.4)',
  black08: 'rgba(0, 0, 0, 0.08)',

  // light mode
  surface: '#FFFFFF',
  background: '#F7F8FA',
  onPrimary: '#FFFFFF',
  backgroundOnPrimary: '#FFFFFF',
  backgroundSurface: '#FFFFFF',
  backgroundOnSurface: '#2D2D2D',
  primary: '#4475E3',
  primaryLight: '#ECF1FC',
  primaryHovered: '#2F65E0',
  primary05: '#F6F8FE',
  primary10: '#ECF1FC',
  primary30: '#C7D6F7',
  primaryDisabled: '#B3C6F1',
  primaryLightPressed: '#E3EAFB',
  primaryLightFocused: '#C7D6F7',
  textPrimary: '#474747',
  primaryWhite: '#FFFFFF',
  disabled: '#DBDBDB',
  disabled20: 'rgba(219, 219, 219, 0.2)',
  onDisabled: '#A1A1A1',

  textPrimaryBlue: '#4475E3',
  textPrimaryDark: '#474747',
  textSecondaryGray: '#808080',
  textDisabled: '#838383',
  onSurface: '#2D2D2D',
  secondarySkyblue: '#00B0D7', // replace
  backgoundLight: '#F0F0F0',
  primaryBlueHovered: '#2F65E0',
  primaryBluePressed: '#2057D5',

  statusSuccess: '#2DC008',
  statusSuccess10: '#DCF2EA',
  statusSuccessText: '#00825D',

  statusError: '#FF3333',
  statusError10: '#FFEBEB',
  statusErrorText: '#D14343',

  statusWarning: '#F5BF00',
  statusWarning10: '#FFF7D0',
  statusWarningText: '#A36400',

  secondaryViolet: '#944ED7',
  secondaryVioletHovered: '#6A3799',
  secondaryViolet10: '#F4EDFB',

  secondaryTangerine: '#F39C18',
  secondaryTangerineHovered: '#CC8314',
  secondaryTangerine10: '#FEF5E8',

  secondarySkyBlue: '#00B0D7',
  secondarySkyBlueHovered: '#0092B3',
  secondarySkyBlue10: '#E5F7FB',

  secondaryGreen: '#1AD598',
  secondaryGreenHovered: '#16B37F',
  secondaryGreen10: '#E8FBF5',

  secondaryRed: '#FA5F4F',
  secondaryRedHovered: '#CC4E40',
  secondaryRed10: '#FFEFED',
};

export type SpecColorType = keyof typeof colors;

const weightToCssValue = {
  regular: 400,
  upperregular: 450,
  medium: 500,
  semibold: 600,
  bold: 700,
};

const fontSpec = (
  size: number,
  lineHeight: number | string,
  weight: keyof typeof weightToCssValue,
  fontFamily: string
) => css`
  font-family: ${fontFamily}, sans-serif;
  font-size: ${px(size)};
  line-height: ${typeof lineHeight === 'string' ? lineHeight : px(lineHeight)};
  font-weight: ${weightToCssValue[weight]};
  font-style: normal;
`;

const typography = {
  // Inter typeface
  labelRegular: fontSpec(10, 15, 'regular', 'Inter'),
  labelSemibold: fontSpec(10, 15, 'semibold', 'Inter'),

  bodyXSmallRegular: fontSpec(12, 18, 'regular', 'Inter'),
  bodyXSmallSemibold: fontSpec(12, 18, 'semibold', 'Inter'),
  bodySmallRegular: fontSpec(14, 21, 'regular', 'Inter'),
  bodySmallSemibold: fontSpec(14, 21, 'semibold', 'Inter'),
  bodyMediumRegular: fontSpec(16, 24, 'regular', 'Inter'),
  bodyMediumSemibold: fontSpec(16, 24, 'semibold', 'Inter'),
  bodyLargeRegular: fontSpec(18, 26, 'regular', 'Inter'),
  bodyLargeSemibold: fontSpec(18, 26, 'semibold', 'Inter'),

  headingXSmall: fontSpec(16, 21, 'semibold', 'Inter'),
  headingSmall: fontSpec(18, 23, 'semibold', 'Inter'),
  headingXMedium: fontSpec(18, 27, 'semibold', 'Inter'),
  headingXMediumRegular: fontSpec(18, 27, 'regular', 'Inter'),
  headingMedium: fontSpec(24, 32.2, 'semibold', 'Inter'),
  headingLargeSemibold: fontSpec(32, 42, 'semibold', 'Inter'),
  headingXLargeSemibold: fontSpec(48, 58, 'semibold', 'Inter'),
  headingXXLargeSemibold: fontSpec(60, 72, 'semibold', 'Inter'),

  // Fira Code typeface
  query14: fontSpec(14, '130%', 'upperregular', 'Fira Code'),
  query12: fontSpec(12, '150%', 'upperregular', 'Fira Code'),
};

const boxShadow = {
  card: `0 0 ${px(2)} ${px(0)} rgba(0,0,0,0.15)`,
  avatar: `0 ${px(8)} ${px(10)} ${px(2)} 0 rgba(0,0,0,0.12)`,
  studyAvatar: `0 ${px(15)} ${px(20)} rgba(180,180,180,0.3)`,
  previewScreen: `0 ${px(2)} ${px(4)} rgba(71,71,71,0.25)`,
};

const animation = {
  defaultTiming: 'cubic-bezier(0.45, 0.05, 0, 1)',
  rippleTiming: 'cubic-bezier(0.4, 0, 0.2, 1)',
};

declare module 'styled-components' {
  export interface DefaultTheme {
    colors: typeof colors;
    typography: typeof typography;
    boxShadow: typeof boxShadow;
    animation: typeof animation;
  }
}

const theme: DefaultTheme = {
  colors,
  typography,
  boxShadow,
  animation,
};

export default theme;
