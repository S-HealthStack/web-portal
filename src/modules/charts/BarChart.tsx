import React, { useRef, useMemo, useState, useCallback } from 'react';
import styled from 'styled-components';
import * as d3 from 'd3';

import Tooltip from 'src/common/components/Tooltip';
import { colors, px, theme, typography } from 'src/styles';
import { XAxis, YAxis, Bar } from './common-components';
import {
  TooltipContent,
  NO_RESPONSES_LABEL,
  MARGIN_FOCUS,
  NoResponsesLabel,
} from './common-helpers';

interface StyledSvgProps {
  $isHorizontal?: boolean;
}

const Container = styled.div`
  position: relative;
`;

const StyledSvg = styled.svg<StyledSvgProps>`
  overflow: visible;

  .xAxis {
    .tick {
      line {
        stroke: ${colors.backgoundLight};
        stroke-width: ${px(1)};
        opacity: ${({ $isHorizontal }) => ($isHorizontal ? 1 : 0)};
      }

      text,
      foreignObject > div {
        ${typography.labelRegular};
        color: ${colors.textPrimary};
      }

      foreignObject > div {
        text-align: center;
        display: -webkit-box;
        -webkit-box-orient: vertical;
        -webkit-line-clamp: 2;
        overflow: hidden;
      }
    }
  }

  .yAxis {
    .tick {
      line {
        stroke: ${colors.backgoundLight};
        stroke-width: ${px(1)};
        opacity: ${({ $isHorizontal }) => ($isHorizontal ? 0 : 1)};
      }
      text {
        ${typography.labelRegular};
        color: ${colors.textPrimary};
        text-anchor: end;
      }
    }
  }
`;

const TooltipNameBlock = styled.div`
  display: flex;
  margin-right: ${px(8)};
  ${typography.labelSemibold};
  color: ${colors.surface};
`;

const TooltipDataBlock = styled.div`
  display: flex;
  flex-direction: column;
  ${typography.labelRegular};
  color: ${colors.surface};
  margin: 0;
`;

interface DataItem {
  name: string;
  value: number;
  totalValue: number;
}

export type BarChartProps = {
  data: DataItem[];
  width: number;
  height: number;
  isHorizontal?: boolean;
};

const MAX_VALUE = 100;
const LABEL_MARGIN = 16;

const getMargins = (isHorizontal = false) =>
  isHorizontal
    ? { top: 4, right: 30, bottom: 37, left: 53 }
    : { top: 4, right: 0, bottom: 23, left: 43 };

const BarChart = ({ width, height, data, isHorizontal = false }: BarChartProps) => {
  const svgRef = useRef<SVGSVGElement>(null);

  const [isSmallText, setIsSmallText] = useState(false);

  const margins = useMemo(() => getMargins(isHorizontal), [isHorizontal]);

  const xRange = useMemo(
    () => [margins.left, width - margins.right],
    [margins.left, margins.right, width]
  );
  const yRange = useMemo(
    () => [height - margins.bottom, margins.top],
    [height, margins.bottom, margins.top]
  );

  // y for vertical view, x for horizontal view (% value scale)
  const scaleLinear = useMemo(
    () =>
      d3
        .scaleLinear()
        .domain([0, MAX_VALUE])
        .nice()
        .range(isHorizontal ? xRange : yRange),
    [isHorizontal, xRange, yRange]
  );

  const domainName = useMemo(() => data.map((d) => d.name), [data]);

  // y for horizontal view, x for vertical view (name scale)
  const scaleBand = useMemo(() => {
    const domain = [...domainName];
    if (isHorizontal) {
      domain.reverse();
    }

    return d3
      .scaleBand()
      .domain(domain)
      .range(isHorizontal ? yRange : xRange)
      .paddingInner(isHorizontal ? 0.34 : 0.37);
  }, [domainName, isHorizontal, xRange, yRange]);

  const getPercentage = (d: DataItem) => (d.totalValue ? (d.value / d.totalValue) * 100 : 0);

  const getBarKeyByIndex = (index: number) => `bar-${index}`;

  const [currentHoveredBar, setCurrentHoveredBar] = useState<number>(-1);

  const tooltipContent = useMemo(() => {
    if (currentHoveredBar < 0) {
      return null;
    }

    const barData = data[currentHoveredBar];

    if (!barData || !barData.totalValue) {
      return null;
    }

    return (
      <TooltipContent>
        <TooltipNameBlock>{`${barData.name.toUpperCase()}:`}</TooltipNameBlock>
        <TooltipDataBlock>{`${barData.value}/${barData.totalValue}`}</TooltipDataBlock>
      </TooltipContent>
    );
  }, [currentHoveredBar, data]);

  const tooltipPosition = useMemo(() => {
    if (isHorizontal) {
      return 't';
    }

    switch (currentHoveredBar) {
      case 0:
        return 'atl';
      case data.length - 1:
        return 'atr';
      default:
        return 't';
    }
  }, [currentHoveredBar, data.length, isHorizontal]);

  const onBarMouseEnter = useCallback(
    (_: React.MouseEvent<SVGPathElement, MouseEvent> | null, index: number) => {
      setCurrentHoveredBar(index);
    },
    []
  );

  const onBarMouseLeave = useCallback(() => {
    setCurrentHoveredBar(-1);
  }, []);

  const isEmptyState = useMemo(() => data.every((d) => d.totalValue === 0), [data]);

  const tooltipContainer = useMemo(
    () =>
      currentHoveredBar < 0
        ? undefined
        : (d3
            .select(svgRef.current)
            .select(`#${getBarKeyByIndex(currentHoveredBar)}`)
            .node() as HTMLElement),
    [currentHoveredBar]
  );

  const handleXAxisCustomCall = useCallback(
    (el: d3.Selection<SVGGElement, unknown, HTMLElement, unknown>) => {
      if (el.select('.tick').selectChildren('text').nodes().length && !isHorizontal) {
        const barWidth = scaleBand.bandwidth();

        const oldNodes = el.selectAll('.tick text').remove().nodes();

        el.selectAll('.tick')
          .append('foreignObject')
          .append('xhtml:div')
          .style('width', px(barWidth))
          .style('height', px(30));

        el.selectAll('.tick foreignObject')
          .nodes()
          .forEach((node, idx) => {
            const n = node as SVGTextContentElement;
            const oldN = oldNodes[idx] as SVGTextContentElement;

            if (oldN) {
              oldN.getAttributeNames().forEach((name) => {
                n.setAttribute(name, oldN.getAttribute(name) || '');
              });

              n.setAttribute('width', px(barWidth));
              n.setAttribute('height', px(30));
              n.setAttribute('x', px(-barWidth / 2));
            }
          });

        el.selectAll('.tick foreignObject div')
          .nodes()
          .forEach((node, idx) => {
            (node as SVGTextContentElement).textContent = (
              oldNodes[idx] as SVGTextContentElement
            ).textContent;
          });
      }
    },
    [isHorizontal, scaleBand]
  );

  return (
    <Container>
      <StyledSvg ref={svgRef} width={width} height={height} $isHorizontal={isHorizontal}>
        <XAxis
          xScale={isHorizontal ? scaleLinear : scaleBand}
          removeDomain
          yOffset={height - margins.bottom}
          yTickOffset={isHorizontal ? 20 : 8}
          xTickOffset={!isHorizontal ? 1 : 0}
          ticks={5}
          tickSize={height - margins.top - margins.bottom}
          tickFormatFn={(d) => (isHorizontal ? `${d}%` : `${d}`)}
          orientation={isHorizontal ? 'top' : 'bottom'}
          isScaleBand={!isHorizontal}
          customCall={handleXAxisCustomCall}
        />
        <YAxis
          yScale={isHorizontal ? scaleBand : scaleLinear}
          xOffset={margins.left}
          removeDomain
          ticks={5}
          tickSize={width - margins.left - margins.right}
          tickFormatFn={(d) => (isHorizontal ? `${d}` : `${d}%`)}
          orientation={isHorizontal ? 'left' : 'right'}
          xTickOffset={-30}
          yTickOffset={isHorizontal ? 2 : 1}
          isScaleBand={isHorizontal}
          customCall={(el: d3.Selection<SVGGElement, unknown, HTMLElement, unknown>) => {
            const textItems = el.selectAll('.tick text');
            const textWidth = textItems
              .nodes()
              .map((item: d3.BaseType) => (item as Element).getBoundingClientRect().width);

            textItems.attr('x', (d, i) => (isHorizontal ? -textWidth[i] - 8 : -8));
          }}
        />
        {!isEmptyState &&
          data.map((d, index) => (
            <Bar
              id={getBarKeyByIndex(index)}
              key={getBarKeyByIndex(index)}
              isHorizontal={isHorizontal}
              rectProps={{
                x: isHorizontal ? margins.left : scaleBand(d.name) || 0,
                y: isHorizontal ? +(scaleBand(d.name) || 0) : scaleLinear(getPercentage(d)),
                height: isHorizontal
                  ? scaleBand.bandwidth()
                  : scaleLinear(0) - scaleLinear(getPercentage(d)),
                width: isHorizontal
                  ? scaleLinear(getPercentage(d)) - scaleLinear(0)
                  : scaleBand.bandwidth(),
              }}
              maxHeight={isHorizontal ? scaleBand.bandwidth() : scaleLinear(0) - scaleLinear(100)}
              maxY={isHorizontal ? +(scaleBand(d.name) || 0) : scaleLinear(100)}
              textProps={{
                x: isHorizontal
                  ? scaleLinear(getPercentage(d)) -
                    (getPercentage(d) > 10 ? scaleLinear(0) : 0) +
                    LABEL_MARGIN
                  : +(scaleBand(d.name) || 0) + scaleBand.bandwidth() / 2,
                y: isHorizontal
                  ? +(scaleBand(d.name) || 0) + scaleBand.bandwidth() / 2
                  : height -
                    margins.bottom -
                    (scaleLinear(0) - scaleLinear(getPercentage(d))) +
                    (getPercentage(d) > 10 ? 20 : -8),
                dy: isHorizontal ? '.35em' : '0', // vertical align middle
                fill: getPercentage(d) > 10 ? theme.colors.primaryWhite : theme.colors.textPrimary,
                text: `${Math.round(getPercentage(d))}%`,
              }}
              isSmallText={isSmallText}
              onMouseEnter={(event) => onBarMouseEnter(event, index)}
              onMouseLeave={onBarMouseLeave}
              setIsSmallText={setIsSmallText}
            />
          ))}
        {isEmptyState && <rect x="0px" y="0px" width={width} height={height} fill="transparent" />}
        {isEmptyState && (
          <NoResponsesLabel
            x="50%"
            y="50%"
            dx={MARGIN_FOCUS.left}
            dy={-MARGIN_FOCUS.top}
            textAnchor="middle"
            dominantBaseline="middle"
            width={width}
            height={height}
          >
            {NO_RESPONSES_LABEL}
          </NoResponsesLabel>
        )}
      </StyledSvg>
      <Tooltip
        content={tooltipContent}
        container={tooltipContainer}
        show={!!tooltipContent && !!tooltipContainer}
        position={tooltipPosition}
        arrow
        static
      />
    </Container>
  );
};

export default BarChart;
