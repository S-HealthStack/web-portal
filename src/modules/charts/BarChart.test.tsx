import React from 'react';
import '@testing-library/jest-dom';
import 'jest-styled-components';
import { act, findByTestId, render } from '@testing-library/react';
import { ThemeProvider } from 'styled-components/';
import { userEvent } from '@storybook/testing-library';

import BarChart from 'src/modules/charts/BarChart';
import theme from '../../styles/theme';

describe('BarChart', () => {
  it('should render', async () => {
    const { baseElement } = render(
      <ThemeProvider theme={theme}>
        <BarChart
          data={[{ name: 'test', value: 10, totalValue: 100 }]}
          width={1000}
          height={1000}
        />
      </ThemeProvider>
    );

    expect(baseElement).toMatchSnapshot();

    const bar = await findByTestId(baseElement, 'bar-0');
    expect(bar).toBeInTheDocument();

    act(() => {
      userEvent.hover(bar);
    });

    act(() => {
      userEvent.unhover(bar);
    });
  });

  it('[NEGATIVE] should render with empty data', () => {
    const { baseElement, rerender } = render(
      <ThemeProvider theme={theme}>
        <BarChart data={[{ name: '', value: 0, totalValue: 0 }]} width={1000} height={1000} />
      </ThemeProvider>
    );

    expect(baseElement).toBeDefined();

    rerender(
      <ThemeProvider theme={theme}>
        <BarChart data={[]} width={1000} height={1000} />
      </ThemeProvider>
    );
    expect(baseElement).toBeDefined();
  });
});
