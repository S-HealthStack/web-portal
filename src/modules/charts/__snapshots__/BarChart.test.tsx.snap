// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BarChart should render 1`] = `
.c6 {
  display: inline-block;
  position: relative;
}

.c2 > .tick > text {
  font-family: Inter,sans-serif;
  font-size: 0.625rem;
  line-height: 0.9375rem;
  font-weight: 400;
  font-style: normal;
}

.c3 > .tick > text {
  font-family: Inter,sans-serif;
  font-size: 0.625rem;
  line-height: 0.9375rem;
  font-weight: 400;
  font-style: normal;
}

.c4 {
  fill: #00B0D7;
}

.c5 {
  font-family: Inter,sans-serif;
  font-size: 1rem;
  line-height: 1.3125rem;
  font-weight: 600;
  font-style: normal;
}

.c5.small-color {
  font-family: Inter,sans-serif;
  font-size: 0.75rem;
  line-height: 1.125rem;
  font-weight: 600;
  font-style: normal;
}

.c5 value {
  fill: #F7F8FA;
  -webkit-dominant-baseline: middle;
  dominant-baseline: middle;
  text-anchor: middle;
}

.c0 {
  position: relative;
}

.c1 {
  overflow: visible;
}

.c1 .xAxis .tick line {
  stroke: #F0F0F0;
  stroke-width: 0.0625rem;
  opacity: 0;
}

.c1 .xAxis .tick text,
.c1 .xAxis .tick foreignObject > div {
  font-family: Inter,sans-serif;
  font-size: 0.625rem;
  line-height: 0.9375rem;
  font-weight: 400;
  font-style: normal;
  color: #474747;
}

.c1 .xAxis .tick foreignObject > div {
  text-align: center;
  display: -webkit-box;
  -webkit-box-orient: vertical;
  -webkit-line-clamp: 2;
  overflow: hidden;
}

.c1 .yAxis .tick line {
  stroke: #F0F0F0;
  stroke-width: 0.0625rem;
  opacity: 1;
}

.c1 .yAxis .tick text {
  font-family: Inter,sans-serif;
  font-size: 0.625rem;
  line-height: 0.9375rem;
  font-weight: 400;
  font-style: normal;
  color: #474747;
  text-anchor: end;
}

<body>
  <div>
    <div
      class="c0"
    >
      <svg
        class="c1"
        height="1000"
        width="1000"
      >
        <g
          class="c2 xAxis undefined"
          fill="none"
          font-family="sans-serif"
          font-size="10"
          text-anchor="middle"
          transform="translate(0, 977)"
        >
          <g
            class="tick"
            opacity="1"
            transform="translate(521.5,0)"
          >
            <line
              stroke="currentColor"
              y2="6"
            />
            <foreignobject
              dy="0.71em"
              fill="currentColor"
              height="1.875rem"
              width="37.681875rem"
              x="-18.8409375rem"
              y="8"
            >
              <div
                style="width: 37.681875rem; height: 1.875rem;"
              />
            </foreignobject>
          </g>
        </g>
        <g
          class="c3 yAxis"
          fill="none"
          font-family="sans-serif"
          font-size="10"
          text-anchor="start"
          transform="translate(43,0)"
        >
          <g
            class="tick"
            opacity="1"
            transform="translate(0,977.5)"
          >
            <line
              stroke="currentColor"
              x2="957"
            />
            <text
              dy="0.32em"
              fill="currentColor"
              x="-8"
              y="1"
            >
              0%
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(0,782.9)"
          >
            <line
              stroke="currentColor"
              x2="957"
            />
            <text
              dy="0.32em"
              fill="currentColor"
              x="-8"
              y="1"
            >
              20%
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(0,588.3)"
          >
            <line
              stroke="currentColor"
              x2="957"
            />
            <text
              dy="0.32em"
              fill="currentColor"
              x="-8"
              y="1"
            >
              40%
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(0,393.7)"
          >
            <line
              stroke="currentColor"
              x2="957"
            />
            <text
              dy="0.32em"
              fill="currentColor"
              x="-8"
              y="1"
            >
              60%
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(0,199.09999999999994)"
          >
            <line
              stroke="currentColor"
              x2="957"
            />
            <text
              dy="0.32em"
              fill="currentColor"
              x="-8"
              y="1"
            >
              80%
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(0,4.5)"
          >
            <line
              stroke="currentColor"
              x2="957"
            />
            <text
              dy="0.32em"
              fill="currentColor"
              x="-8"
              y="1"
            >
              100%
            </text>
          </g>
        </g>
        <g
          class="bar"
          data-testid="bar-0"
        >
          <g
            id="bar-0"
          >
            <rect
              class="c4"
              height="97.29999999999995"
              width="602.91"
              x="220.04500000000002"
              y="879.7"
            />
            <text
              class="c5"
              dy="0"
              fill="#474747"
              transform="translate(0)"
              x="521.5"
              y="871.7"
            >
              10%
            </text>
          </g>
          <rect
            fill="transparent"
            height="973"
            width="602.91"
            x="220.04500000000002"
            y="4"
          />
        </g>
      </svg>
      <span
        class="c6"
        data-testid="tooltip-container"
      />
    </div>
  </div>
</body>
`;
